libs/vkd3d/swapchain.c

```diff
@@ -414,26 +450,35 @@ static ULONG STDMETHODCALLTYPE dxgi_vk_swap_chain_AddRef(IDXGIVkSwapChain *iface
 {
     struct dxgi_vk_swap_chain *chain = impl_from_IDXGIVkSwapChain(iface);
     UINT refcount = InterlockedIncrement(&chain->refcount);
+
     TRACE("iface %p, refcount %u\n", iface, refcount);
+
+    if (refcount == 1)
+        dxgi_vk_swap_chain_incref(chain);
+
     return refcount;
 }
 
 static ULONG STDMETHODCALLTYPE dxgi_vk_swap_chain_Release(IDXGIVkSwapChain *iface)
 {
     struct dxgi_vk_swap_chain *chain = impl_from_IDXGIVkSwapChain(iface);
-    struct d3d12_command_queue *queue = chain->queue;
     UINT refcount;
 
     refcount = InterlockedDecrement(&chain->refcount);
     TRACE("iface %p, refcount %u\n", iface, refcount);
 
     if (!refcount)
     {
+        /* Calling this from the submission thread will result in a deadlock, so
+         * drain the swapchain queue now. */
         dxgi_vk_swap_chain_drain_queue(chain);
-        dxgi_vk_swap_chain_cleanup(chain);
-        vkd3d_free(chain);
-        ID3D12CommandQueue_Release(&queue->ID3D12CommandQueue_iface);
+
+        if (chain->queue->device->vk_info.NV_low_latency2)
+            d3d12_device_remove_swapchain(chain->queue->device, chain);
+
+        dxgi_vk_swap_chain_decref(chain);
```

Comment:
Need to publicly decref the ID3D12Device when public ref-count hits 0. Similarly on incref, then reacquire public device refcount. See d3d12_pipeline_state_Release() as an example. d3d12_device_release() must be called after private decref.

Reply:
I have added device refs for AddRef, and Release for the IDXGIVkSwapChain implementation.

Reply:
Actually, I brain-farted I think. It's not the d3d12_device we need to have a public ref on, it's the ID3D12CommandQueue. dxgi_vk_swap_chain_AddRef() may need to reacquire a ref on the command queue, not device. The ID3D12CommandQueue_Release() call here is appropriate as-is.

Reply:
ID3D12CommandQueue_Release(&queue->ID3D12CommandQueue_iface); should still be done here.